src/bin/main.rs

```diff
@@ -305,51 +298,116 @@ fn main() {
     if let Some(code_dir) = code_dir {
-        if let Some(code_files) = code_files {
-            for code_file in &code_files {
-                eprintln!("Copying code file {:?}", code_file);
-                let mut file_path = code_dir.clone();
-                file_path.push(code_file);
-                fs::create_dir_all(file_path.parent().unwrap()).unwrap();
-                match fs::copy(&code_file, &file_path) {
-                    Ok(_) => {}
-                    Err(err) => eprintln!("  --> Error copying code file {:?}: {}", code_file, err),
+        if let Some(code_file_patterns) = &paths.code_files {
+            for (idx, code_file_pattern) in code_file_patterns.iter().enumerate() {
+                let code_path = paths.code_paths.as_ref().map(|code_paths| &code_paths[idx]);
+                for code_file in PathUtil::list_files(code_file_pattern).unwrap() {
+                    let out_path = code_path.map_or(code_file.clone(), |code_path| {
+                        modify_path(&code_file, &code_path)
+                    });
+                    match track_copy_dest.entry(out_path.clone()) {
+                        Occupied(entry) => {
+                            eprintln!(
+                                "ERROR: Attempted to copy multiple code files to {}: from {} and {}",
+                                out_path.display(), entry.get().display(), code_file.display()
+                            );
+                            return;
+                        }
+                        Vacant(entry) => {
+                            entry.insert(code_file.clone());
+                        }
+                    }
+                    eprintln!(
+                        "Copying code file {} to {}",
+                        code_file.display(),
+                        out_path.display()
+                    );
+
+                    let mut file_path = code_dir.clone();
+                    file_path.push(out_path);
+
+                    fs::create_dir_all(file_path.parent().unwrap()).unwrap();
+                    match fs::copy(&code_file, &file_path) {
+                        Ok(_) => {}
+                        Err(err) => eprintln!(
+                            "ERROR: --> Error copying code file {}: {}",
+                            code_file.display(),
+                            err
+                        ),
+                    }
                 }
             }
         }
     }
+
     if let Some(doc_dir) = doc_dir {
-        if let Some(doc_files) = doc_files {
-            for doc_file in &doc_files {
-                eprintln!("Copying doc file {:?}", doc_file);
+        if let Some(doc_files) = &paths.doc_files {
+            for doc_file in PathUtil::list_all_files(doc_files).unwrap() {
+                eprintln!("Copying doc file {}", doc_file.display());
                 let mut file_path = doc_dir.clone();
-                file_path.push(doc_file);
+                file_path.push(&doc_file);
```

Comment:
This seems to be taking a ref of a ref now?

Reply:
No, doc_file is `PathBuf`. `PathUtil::list_all_files(doc_files)` returns `Result<Vec<PathBuf>`.